src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -932,12 +964,16 @@ data Knowledge = KN
   , knRWTerminationOpts :: RWTerminationOpts
   }
 
-isValid :: Knowledge -> Expr -> IO Bool
-isValid γ e = do
-  contra <- knPreds γ (knContext γ) (knLams γ) PFalse
-  if contra
-    then return False
-    else knPreds γ (knContext γ) (knLams γ) e
+isValid :: IORef (M.HashMap Expr Bool) -> Knowledge -> Expr -> IO Bool
+isValid cacheRef γ e = do
+    smtCache <- readIORef cacheRef
+    case M.lookup e smtCache of
+      Nothing -> do
+        b <- knPreds γ (knContext γ) (knLams γ) e
+        when b $
+          writeIORef cacheRef (M.insert e True smtCache)
```

Comment:
I am puzzled by one thing: the `e := True` is _written_  to the cache depending on the result at a _particular_ environment `γ`. However,
it is then later _read_ from the cache independent of the environment.

Why is this sound? 

For example, in a particular `γ` we may have `x == y` which will tell us that the two terms are equal (under `γ`) but the terms may not be equal in _other_ contexts? 

Perhaps I'm missing something?

Reply:
Perhaps there is unsoundness. In my analysis of the code, it looks like `γ` is never trimmed while the cached smt queries are kept.

If `γ` is trimmed when switching sibling nodes of the Trie, it wouldn't be a problem because the cache isn't shared across siblings.

Said otherwise, let's look for unsoundness here. When are equalities retired from the cache `γ`?

Reply:
Ah, so equalities are never retired as such, but `γ` only "grows" according to the call stack. [I wonder if I can draw a picture.]

A constraint that looks like (with equalities in `p1`, `p2`, `p3` ...) 

```
forall x1:p1 => 
      forall x2:p2 => q
  /\  forall x3:p3 => r
```

Generates a trie like 

```
x1:p1
`------ x2:p2 
|             `----------- q
|            
`------ x3:p3
             `------------ r
```            

And the (old) PLE traverses the trie tracking the environment
explicitly as a parameter with `[]`, then `[x1:p1]` then `[x1:p1, x2:p2]` *then back to* `[x1:p1]` then `[x1:p1, x3:p3]` ...

It seemed to me that your `cache` is now "global" so the equalities from `p2` would also be in the context when we're at `r` ... 

However, you write that

> because the cache isn't shared across siblings.

Can you clarify why that happens?

Reply:
[In particular, previously the `ICtx` holds the learned information at "each node" of the trie, and we explicitly pass it around so that we have "different" `ICtx` at different siblings]

Reply:
>> because the cache isn't shared across siblings.
>
> Can you clarify why that happens?

At the point in the history that this change is introduced (5c8beb3), every call to `evalCandsLoop` starts with an empty cache (which comes in the `env0` parameter). This means the cache is only used to discover the equalities of a single node in the trie.

This is changed later on in 75c49b16, where the cache is passed to all of the descendents of a node. However, on each path of the trie, different copies of the cache are used, so queries added to one copy don't affect the others.

Reply:
> [In particular, previously the ICtx holds the learned information at "each node" of the trie, and we explicitly pass it around so that we have "different" ICtx at different siblings]

I'm fuzzy still at whether to put some fields in `ICtx` or `EvalEnv`. In 8f942ba5 I'm removing `icFuel` for instance, which the code struggled to keep in sync between both `ICtx` and `EvalEnv`.

Perhaps we could reserve `ICtx` for cross-node information and `EvalEnv` for node-local information, but then it puzzles me that an `EvalEnv` is provided from outside `evalCandsLoop` as a parameter :).

Reply:
Yes you’re right - but let’s merge these in and then we can look into that separately?